File: examples/typescript/src/spine/pure/verb.ts
// ── spine/pure/verb — a registration, and what a tool may read (6.8, G1) ────
// ONE TOOL MECHANIC. 6.8's "a UI tool folds, does not sign" carve-out is gone.
// A presentation verb (focusTicket, setPanel) declares exactly what a domain
// verb declares — name, description, input schema, pure run, sign, and its
// reversibility classification — and costs exactly the same four appends. There
// is no second table and no cheaper UI path, because there is no UI path.
//
// `Verb` is SEALED (Reversible | Irreversible) and that is what makes 14.3's
// DEFAULT-DENY structural: there is no default. You pick a variant, and
// `Irreversible` cannot be constructed without saying where its matching
// Request lives (`requestedBy`). Registering a tool FORCES the classification
// decision, and the classification is a reviewed row in the registry.

import type { Authority, Signature } from "./actor";
import type { CommandBase } from "./command";
import type { Context } from "./context";
import type { Attributed, EffectBase } from "./effect";
import type { CommandId, RawInput, Timestamp, ToolName } from "./ids";
import type { Notice } from "./notice";
import type { StagedInput } from "./staged";
import type { ToolResultBase } from "./tool-result";

// ── What a tool may read ────────────────────────────────────────────────────
// NO actor. NO authority. NO Signature. Deleting `ctx.actor` is what makes an
// Actor UNREPRESENTABLE upstream of the boundary (G1): a tool asking "who is
// asking?" is asking the wrong question, because the answer is stamped after it
// returns. The two-unreconciled-actor-values bug cannot recur, because there is
// only one value and it is created after the tool has returned.
export interface Ctx<S> {
  /** the committed snapshot, read-only */
  readonly state: S;
  /** the bounded projection the reasoner also saw (G15) */
  readonly context: Context;
}

// ── Input decoding ──────────────────────────────────────────────────────────
// The spine names a STANDARD, not a library.
//
// This used to declare `safeParse(raw) => { success, data }` and claim to be
// library-neutral because the shape was structural. It was not: that is zod's method
// and zod's field names. Valibot returns `{ success, output }` from a STANDALONE
// `safeParse(schema, input)`, so it could not satisfy the interface at all — the
// neutrality was a comment, not a property.
//
// Standard Schema (https://standardschema.dev) is the vendor-neutral contract both
// implement: a `~standard.validate` that returns `{ value }` on success and
// `{ issues }` on failure. Zod v4, Valibot v1 and ArkType all ship it, and the Vercel
// AI SDK accepts it directly for the model-facing tool definition — so ONE object
// serves both the reasoner's schema and the boundary's decoder, which is what 6.8
// requires.
export interface StandardResult<I> {
  readonly value?: I;
  readonly issues?: readonly unknown[];
}

export interface InputSchema<I> {
  readonly "~standard": {
    readonly validate: (value: unknown) => StandardResult<I> | Promise<StandardResult<I>>;
  };
}

export type DecodeResult = { readonly ok: true; readonly input: unknown } | { readonly ok: false };

// ── The typed spec a block writes ───────────────────────────────────────────
export interface VerbSpec<S, I, R extends ToolResultBase, C extends CommandBase> {
  /** 6.8 — the tool name IS the result's discriminant and the registry key */
  readonly name: R["tool"] & ToolName;
  /** the model-facing description */
  readonly describe: string;
  /** the input schema — model-facing AND the boundary's decoder */
  readonly schema: InputSchema<I>;
  /** the PURE tool body: reads ctx, returns a payload, mutates nothing */
  readonly run: (input: I, ctx: Ctx<S>) => R;
  /** the name→Command entry (6.8) */
  readonly sign: (result: R, sig: Signature, id: CommandId) => C;
}

// ── The type-erased registry entry ──────────────────────────────────────────
// The spine holds verbs for many different input and result types in one map,
// so the registry entry is erased to the shared base types. The two casts that
// erasure needs live HERE, in one file, and nowhere else in the system.
export interface VerbBase<S> {
  readonly kind: "Reversible" | "Irreversible";
  readonly name: ToolName;
  readonly describe: string;
  /** opaque to the spine; only the model-facing adapter interprets it */
  readonly schema: unknown;
  readonly decode: (raw: RawInput) => DecodeResult;
  readonly run: (input: unknown, ctx: Ctx<S>) => ToolResultBase;
  readonly sign: (result: ToolResultBase, sig: Signature, id: CommandId) => CommandBase;
}

export interface ReversibleVerb<S> extends VerbBase<S> {
  readonly kind: "Reversible";
}

export interface IrreversibleVerb<S> extends VerbBase<S> {
  readonly kind: "Irreversible";
  /** where the matching Request lives — the authority that ASKED, read from
   *  committed State. `null` means "no pending request", which the gate refuses. */
  readonly requestedBy: (state: S, result: ToolResultBase) => Authority | null;
}

export type Verb<S> = ReversibleVerb<S> | IrreversibleVerb<S>;

function erase<S, I, R extends ToolResultBase, C extends CommandBase>(
  spec: VerbSpec<S, I, R, C>,
): Omit<VerbBase<S>, "kind"> {
  return {
    name: spec.name,
    describe: spec.describe,
    schema: spec.schema,
    decode: (raw) => {
      const parsed = spec.schema["~standard"].validate(raw);
      // A Standard Schema MAY validate asynchronously. Decoding runs inside the pure
      // fold path, which cannot await, so an async schema is a decode failure rather
      // than a silently-unresolved promise treated as a value.
      if (parsed instanceof Promise) return { ok: false };
      return parsed.issues === undefined && parsed.value !== undefined
        ? { ok: true, input: parsed.value }
        : { ok: false };
    },
    run: (input, ctx) => spec.run(input as I, ctx),
    sign: (result, sig, id) => spec.sign(result as R, sig, id),
  };
}

export function reversible<S, I, R extends ToolResultBase, C extends CommandBase>(
  spec: VerbSpec<S, I, R, C>,
): ReversibleVerb<S> {
  return { kind: "Reversible", ...erase(spec) };
}

export function irreversible<S, I, R extends ToolResultBase, C extends CommandBase>(
  spec: VerbSpec<S, I, R, C> & { readonly requestedBy: (state: S, result: R) => Authority | null },
): IrreversibleVerb<S> {
  return {
    kind: "Irreversible",
    ...erase(spec),
    requestedBy: (state, result) => spec.requestedBy(state, result as R),
  };
}

// ── What a fold arm returns (12.4) ──────────────────────────────────────────
// Three rules, mechanical, no exceptions:
//   1. every arm reads current state before it decides;
//   2. every effect push lives INSIDE the success branch;
//   3. a rejection folds a per-item Notice, never the session-global RunStatus.
export interface ArmOut<S> {
  readonly slice: S;
  readonly effects: readonly EffectBase[];
  readonly notices: readonly Notice[];
}

export function armOut<S>(
  slice: S,
  effects: readonly EffectBase[] = [],
  notices: readonly Notice[] = [],
): ArmOut<S> {
  return { slice, effects, notices };
}

// ── What the root contributes back to the spine ─────────────────────────────
// `effects` is ATTRIBUTED, not bare. An `Attributed` has no `kind` and no `at`,
// so it is not an `EffectBase` and `keyedEffect(step, i, …)` will not take one:
// the only route from what the fold returned to what the sink performs is
// `admit` (spine/pure/effect.ts), which is what makes the refusal a property of
// the DERIVATION rather than of the live path (docs/DECISIONS.md:85).
export interface FoldOut<S> {
  readonly state: S;
  readonly effects: readonly Attributed[];
}

/** The two total dispatchers the impure seam needs. Declared with METHOD syntax
 *  deliberately: the app implements them over its own CLOSED unions
 *  (`readonly ToolResult[]`), while the spine may only name the open base. */
export interface Dispatchers<S> {
  fold(state: S, results: readonly ToolResultBase[], now: Timestamp, sig: Signature): FoldOut<S>;
  projectContext(state: S, staged: readonly StagedInput[]): Context;
}

/** A block's ONE public contribution to the composition root (G11). */
export interface BlockRegistration<S> {
  readonly block: string;
  readonly verbs: readonly Verb<S>[];
}

File: examples/kotlin/src/main/kotlin/adr/spine/pure/Verb.kt
// ── spine/pure/verb — one registration, two maps, default-deny by construction ─
// A Verb is everything a block declares about one tool: its name, the description
// the model reads, the input schema, the PURE body, the name→Command entry, and
// its reversibility.
//
// 14.3's default-deny becomes structural: there is no default. You pick a variant,
// and Irreversible CANNOT BE CONSTRUCTED without saying where its matching Request
// lives. Registering a tool FORCES the classification decision, and the
// classification is a reviewed row in the registry, not a guess.
//
// A1: there is exactly ONE tool mechanic. A presentation verb (focusTicket,
// setPanel) is a Verb with the same five properties, the same signer and the same
// blast radius as a domain verb (setPriority). There is no second table.

package adr.spine.pure

import adr.contract.Command
import adr.contract.ToolResult

/**
 * What a tool may read: the committed snapshot, and the bounded projection the
 * reasoner also saw.
 *
 * NO actor. NO authority. NO Signature (F2/D4). Deleting ctx.actor is what makes an
 * Actor UNREPRESENTABLE upstream of the boundary — a tool asking "who is asking?"
 * is asking the wrong question, because the answer is stamped after it returns.
 */
data class Ctx<S>(val state: S, val context: Context)

/**
 * What the boundary gate must decide about a verb — computed by the verb itself
 * from its OWN classification, because Kotlin's generic erasure means the boundary
 * cannot `is`-check `Verb.Irreversible<S, *, *>` and still call through it.
 * The sealed pair is what the gate matches on, exhaustively.
 */
sealed interface Gating {
    /** A Reversible verb: the gate passes it through untouched. */
    data object Ungated : Gating

    /** An Irreversible verb: it proceeds only if a DIFFERENT authority requested it. */
    data class NeedsConfirmation(val requestedBy: Authority?) : Gating
}

sealed interface Verb<S, I, R : ToolResult> {
    /** The verb name — the discriminant of its ToolResult, of its Command, and the registry key (D3). */
    val name: ToolName

    /** The model-facing description. The only prose the reasoner is given about this tool. */
    val describe: String

    /** The input schema: raw input in, typed input or null out. */
    val decode: (RawInput) -> I?

    /** The PURE tool body. Reads Ctx; returns a payload; mutates nothing (G2). */
    val run: (I, Ctx<S>) -> R

    /** The name→Command entry (6.8). Every verb signs — domain and presentation alike (A1). */
    val sign: (R, Signature, CommandId) -> Command

    /** decode ∘ run. Returns null when the input failed to decode — the boundary owns that word. */
    fun resolve(input: RawInput, ctx: Ctx<S>): ToolResult? = decode(input)?.let { run(it, ctx) }

    /**
     * Sign a result whose static type has been erased by the registry's star projection.
     *
     * This is the ONE unchecked cast in the system, and D3 is its proof: a result
     * reaches this verb only because its `tool` name looked this verb up, and one
     * name means one result case. Gate check C13 re-proves that mapping mechanically
     * for every case in the system, so the cast is total by test, not by luck.
     */
    @Suppress("UNCHECKED_CAST")
    fun signOf(result: ToolResult, sig: Signature, id: CommandId): Command =
        sign(result as R, sig, id)

    /** The verb's own answer to "must the boundary gate this?" */
    fun gating(state: S, result: ToolResult): Gating

    /** Undoable, or cheap to undo. Ungated. */
    data class Reversible<S, I, R : ToolResult>(
        override val name: ToolName,
        override val describe: String,
        override val decode: (RawInput) -> I?,
        override val run: (I, Ctx<S>) -> R,
        override val sign: (R, Signature, CommandId) -> Command,
    ) : Verb<S, I, R> {
        override fun gating(state: S, result: ToolResult): Gating = Gating.Ungated
    }

    /**
     * Not undoable. Cannot be constructed without `requestedBy`: the lens that reads,
     * out of committed State, WHICH AUTHORITY asked for this — because 14.3 requires
     * the confirmation to come from a different principal than the one that asked.
     */
    data class Irreversible<S, I, R : ToolResult>(
        override val name: ToolName,
        override val describe: String,
        override val decode: (RawInput) -> I?,
        override val run: (I, Ctx<S>) -> R,
        override val sign: (R, Signature, CommandId) -> Command,
        val requestedBy: (S, ToolResult) -> Authority?,
    ) : Verb<S, I, R> {
        override fun gating(state: S, result: ToolResult): Gating =
            Gating.NeedsConfirmation(requestedBy(state, result))
    }
}

/** The one word the system uses when an input did not decode. Spelled once (F1). */
const val DECODE_FAILED = "input failed to decode"

/**
 * What the MODEL is told about one call, so it has a payload to reason over.
 *
 * The RECORDED truth is produced separately, at the boundary (§3.1, §15 risk 4);
 * this string never folds, never signs and never reaches the timeline.
 *
 * It lives HERE and not in spine/agent/loop because it makes a decision — what to
 * say when the input did not decode — and G3 says the loop is a declaration, not a
 * place for policy. Gate check C14 denies the branch the moment it drifts back into
 * the loop file, which is exactly how it got there in the first place.
 */
fun <S> Verb<S, *, *>.modelEcho(input: RawInput, ctx: Ctx<S>): String =
    resolve(input, ctx)?.toString() ?: DECODE_FAILED

/** The one closed table the boundary reads: name → verb. It supplies BOTH maps (F1). */
typealias Registry<S> = Map<ToolName, Verb<S, *, *>>

/** A block's single public contribution to the spine (I7). */
data class BlockRegistration<S>(val block: String, val verbs: List<Verb<S, *, *>>)

/** The composition root's one call: registrations in, the closed registry out. */
fun <S> registryOf(vararg blocks: BlockRegistration<S>): Registry<S> {
    val verbs = blocks.flatMap { it.verbs }
    val registry = verbs.associateBy { it.name }
    check(registry.size == verbs.size) { "two blocks registered the same tool name" }
    return registry
}
